packages/http-client/src/index.ts

```diff
@@ -1,80 +1,129 @@
-    include.breakpoints && (await fetch(breakpointsUrl)).json(),
-    include.pages && (await fetch(pagesUrl)).json(),
-  ]);
+}): Promise<Project | string> => {
+  try {
+    if (apiUrl === undefined) {
+      throw new Error("Webstudio API URL is required.");
+    }
+    const urls = [];
+    const gatheredPages: Project = {};
 
-  if (include.pages && listOfPages.pages && listOfPages.pages.length > 0) {
-    for (const page of listOfPages.pages) {
+    const baseUrl = new URL(apiUrl);
+    const pagesUrl = new URL(`/rest/pages/${projectId}`, baseUrl);
+    const listOfPages = await (await fetch(pagesUrl)).json();
+    if (listOfPages === undefined || listOfPages.errors) {
+      throw new Error(`Project ${projectId} needs to be published first`);
+    }
+    const { pages, homePage } = listOfPages;
+    if (homePage) {
+      const { id } = homePage;
+      const treeUrl = new URL(`/rest/tree/${projectId}/${id}`, baseUrl);
+      const propsUrl = new URL(`/rest/props/${projectId}/${id}`, baseUrl);
+      const breakpointsUrl = new URL(`/rest/breakpoints/${projectId}`, baseUrl);
+      const cssUrl = new URL(`/rest/css/${projectId}/${id}`, baseUrl);
       urls.push({
-        url: new URL(`/rest/tree/${projectId}/${page.id}`, baseUrl),
-        path: page.path,
+        page: { ...homePage },
+        url: treeUrl,
+        path: "/",
         type: "tree",
       });
       urls.push({
-        url: new URL(`/rest/props/${projectId}/${page.id}`, baseUrl),
-        path: page.path,
+        page: { ...homePage },
+        url: propsUrl,
+        path: "/",
         type: "props",
       });
+      urls.push({
+        page: { ...homePage },
+        url: breakpointsUrl,
+        path: "/",
+        type: "breakpoints",
+      });
+      urls.push({
+        page: { ...homePage },
+        url: cssUrl,
+        path: "/",
+        type: "css",
+      });
     }
-    if (urls.length === 0) {
-      return {
-        tree,
-        props,
-        breakpoints,
-        pages,
-      };
+
+    if (pages && pages.length > 0) {
+      for (const page of pages) {
+        const { id, path } = page;
+        const treeUrl = new URL(`/rest/tree/${projectId}/${id}`, baseUrl);
+        const propsUrl = new URL(`/rest/props/${projectId}/${id}`, baseUrl);
+        const cssUrl = new URL(`/rest/css/${projectId}/${id}`, baseUrl);
+
+        urls.push({
+          page,
+          url: treeUrl,
+          path,
+          type: "tree",
+        });
+
+        urls.push({
+          page,
+          url: propsUrl,
+          path,
+          type: "props",
+        });
+
+        urls.push({
+          page,
+          url: cssUrl,
+          path,
+          type: "css",
+        });
+      }
     }
 
-    const pagesData = await Promise.all(
+    if (urls.length === 0) {
+      throw new Error("No pages to fetch");
+    }
+    const data = await Promise.all(
       urls.map(async (url) => {
         const res = await fetch(url.url);
         if (res.ok) {
```

Comment:
I am not quite sure but I think you are serially awaiting here, not fetching in parallel

Reply:
Looks like parallel to me, but it actually can be a bad idea to send all these requests at once. If we assume there can be 100s of pages, we might be sending 300 requests at once here.

Not sure it should be a priority now, but we should probably implement a queue where only a fixed number of requests is handled in parallel. Maybe there's a library for that.

I'd actually consider using https://github.com/graphql/dataloader , as it would allow to write code as if we're fetching everything serially instead of building a `urls` array, but under the hood it would send requests in parallel. This would make code more straightforward and with better TypeScript types. And it has a `maxBatchSize` setting.

Reply:
After thinking more: probably a bad idea with dataloader actually. Not a good fit for the problem

Reply:
Yes, that is a valid problem (100s pages). 
A proper solution would be to use a single endpoint dedicated to fetching all project details and sending a response.

Reply:
I go this way.

Reply:
there are some tiny abstractions out there that allow you to nicely limit the parallelization level, e.g. limit it to 20 pages in parallel or something ... even if you add an endpoint that does all the work on the node server, is it actually making it faster? This would only be the case if you loading more data from the database for each request than you need.

Reply:
It is not about how fast data loads but the number of requests going back and forth. With the single endpoint, it is only one.